ignite/services/plugin/plugin.go

```diff
@@ -0,0 +1,338 @@
+func Update(plugins ...*Plugin) error {
+	for _, p := range plugins {
+		err := p.clean()
+		if err != nil {
+			return err
+		}
+		p.fetch()
+	}
+	return nil
+}
+
+// newPlugin creates a Plugin from configuration.
+func newPlugin(pluginsDir string, cp chainconfig.Plugin) *Plugin {
+	var (
+		p          = &Plugin{Plugin: cp}
+		pluginPath = cp.Path
+	)
+	if pluginPath == "" {
+		p.Error = errors.Errorf(`missing plugin property "path"`)
+		return p
+	}
+	if pluginPath[0] == '/' {
+		// This is a local plugin, check if the file exists
+		st, err := os.Stat(pluginPath)
+		if err != nil {
+			p.Error = errors.Wrapf(err, "local plugin path %q not found", pluginPath)
+			return p
+		}
+		if !st.IsDir() {
+			p.Error = errors.Errorf("local plugin path %q is not a dir", pluginPath)
+			return p
+		}
+		p.srcPath = pluginPath
+		p.binaryName = path.Base(pluginPath)
+		return p
+	}
+	// This is a remote plugin, parse the URL
+	if i := strings.LastIndex(pluginPath, "@"); i != -1 {
+		// path contains a reference
+		p.reference = pluginPath[i+1:]
+		pluginPath = pluginPath[:i]
+	}
+	parts := strings.Split(pluginPath, "/")
+	if len(parts) < 3 {
+		p.Error = errors.Errorf("plugin path %q is not a valid repository URL", pluginPath)
+		return p
+	}
+	p.repoPath = path.Join(parts[:3]...)
+	p.cloneURL = "https://" + p.repoPath
+	if len(p.reference) > 0 {
+		p.repoPath += "@" + p.reference
+	}
+	p.cloneDir = path.Join(pluginsDir, p.repoPath)
+	p.srcPath = path.Join(pluginsDir, p.repoPath, path.Join(parts[3:]...))
+	p.binaryName = path.Base(pluginPath)
+	return p
+}
+
+func (p *Plugin) isLocal() bool {
+	return p.cloneURL == ""
+}
+
+func (p *Plugin) binaryPath() string {
+	return path.Join(p.srcPath, p.binaryName)
+}
+
+// load tries to fill p.Interface, ensuring the plugin is usable.
+func (p *Plugin) load() {
+	if p.Error != nil {
+		return
+	}
+	_, err := os.Stat(p.srcPath)
+	if err != nil {
+		// srcPath found, need to fetch the plugin
+		p.fetch()
+		if p.Error != nil {
+			return
+		}
+	}
+	if p.isLocal() {
+		// trigger rebuild for local plugin if binary is outdated
+		if p.outdatedBinary() {
+			p.build()
+		}
+	} else {
+		// Check if binary is already build
+		_, err = os.Stat(p.binaryPath())
+		if err != nil {
+			// binary not found, need to build it
+			p.build()
+		}
+	}
+	if p.Error != nil {
+		return
+	}
+	// pluginMap is the map of plugins we can dispense.
+	pluginMap := map[string]hplugin.Plugin{
+		p.binaryName: &InterfacePlugin{},
+	}
+	// Create an hclog.Logger
+	logger := hclog.New(&hclog.LoggerOptions{
+		Name:   fmt.Sprintf("plugin %s", p.Name),
+		Output: os.Stdout,
+	})
+	// We're a host! Start by launching the plugin process.
+	client := hplugin.NewClient(&hplugin.ClientConfig{
+		HandshakeConfig: handshakeConfig,
+		Plugins:         pluginMap,
+		Logger:          logger,
+		Cmd:             exec.Command(p.binaryPath()),
+		SyncStderr:      os.Stderr,
+		SyncStdout:      os.Stdout,
+	})
+	// defer client.Kill()
```

Comment:
Wouldn't it be good to keep a reference to the client in the `Plugin` so the `client.Kill()` function could be called before the Cobra command finishes to stop and cleanup the plugin processes that are running in the background?

Reply:
Good idea, I had to change the plugin logger level from `info` to `error` because the `client.Kill()` outputs logs, for instance :

```
$ ignite  print
Account list [{alice [20000token 200000000stake]    } {bob [10000token 100000000stake]    }]
2022-10-10T17:51:05.945+0200 [INFO]  plugin print: plugin process exited: path=/home/tom/.ignite/plugins/git.sr.ht/~tbruyelle/plugins-sp/print/print pid=140120
```

9c1858cb